test/Microsoft.AspNet.Cors.Test/CorsTestFixture.cs

```diff
@@ -0,0 +1,145 @@
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Reflection;
+using System.Runtime.Versioning;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Hosting;
+using Microsoft.AspNet.TestHost;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Testing;
+using Microsoft.Extensions.PlatformAbstractions;
+
+namespace Microsoft.AspNet.Cors.Infrastructure
+{
+    public class CorsTestFixture : IDisposable
+    {
+        private readonly TestServer _server;
+
+        public CorsTestFixture(object startupInstance)
+        {
+            var startupTypeInfo = startupInstance.GetType().GetTypeInfo();
+            var configureApplication = (Action<IApplicationBuilder>)startupTypeInfo
+                .DeclaredMethods
+                .FirstOrDefault(m => m.Name == "Configure" && m.GetParameters().Length == 1)
+                ?.CreateDelegate(typeof(Action<IApplicationBuilder>), startupInstance);
+            if (configureApplication == null)
+            {
+                var configureWithLogger = (Action<IApplicationBuilder, ILoggerFactory>)startupTypeInfo
+                    .DeclaredMethods
+                    .FirstOrDefault(m => m.Name == "Configure" && m.GetParameters().Length == 2)
+                    ?.CreateDelegate(typeof(Action<IApplicationBuilder, ILoggerFactory>), startupInstance);
+                Debug.Assert(configureWithLogger != null);
+
+                configureApplication = application => configureWithLogger(application, NullLoggerFactory.Instance);
+            }
+
+            var buildServices = (Func<IServiceCollection, IServiceProvider>)startupTypeInfo
+                .DeclaredMethods
+                .FirstOrDefault(m => m.Name == "ConfigureServices" && m.ReturnType == typeof(IServiceProvider))
+                ?.CreateDelegate(typeof(Func<IServiceCollection, IServiceProvider>), startupInstance);
+            if (buildServices == null)
+            {
+                var configureServices = (Action<IServiceCollection>)startupTypeInfo
+                    .DeclaredMethods
+                    .FirstOrDefault(m => m.Name == "ConfigureServices" && m.ReturnType == typeof(void))
+                    ?.CreateDelegate(typeof(Action<IServiceCollection>), startupInstance);
+                Debug.Assert(configureServices != null);
+
+                buildServices = services =>
+                {
+                    configureServices(services);
+                    return services.BuildServiceProvider();
+                };
+            }
+
+            _server = TestServer.Create(
+                configureApplication,
+                configureServices: InitializeServices(startupTypeInfo.Assembly, buildServices));
+
+            Client = _server.CreateClient();
+            Client.BaseAddress = new Uri("http://localhost");
```

Comment:
> What is "standard"? I copied this straight from the MVC repo, changing it only to remove some MVC-specific service additions.

`TestHost` is a package we want our users to be able to use. We should be working to address the issues that block us from using it with MVC.

I think it would be great if we didn't copy-paste the code from MVC, especially where the extra things that it does aren't required.